### ISSUE TITLE
Microsoft.AppCenter.Storage.StorageException: Failed to prepare SQL query, result=5 database is locked

### ISSUE THREAD

**Issue by tomaustin700** [microsoft/appcenter-sdk-dotnet#1423]
### **Description**
Hi,
We are seeing the following error reported into App Center from our clients. We are not able to duplicate it locally however it does seem to be happening for our clients fairly often.

Fix Microsoft.AppCenter.Storage.StorageException in <>c__DisplayClass31_0`1.<AddTaskToQueue>b__0 ()
Microsoft.AppCenter.Storage.StorageException: Failed to prepare SQL query, result=5 database is locked

Microsoft.AppCenter.Storage.Storage
<>c__DisplayClass31_0`1.<AddTaskToQueue>b__0 ()
System.Threading.Tasks
Task`1.InnerInvoke ()
System.Threading.Tasks
Task.Execute ()
System.Runtime.ExceptionServices
ExceptionDispatchInfo.Throw ()
System.Runtime.CompilerServices
TaskAwaiter.HandleNonSuccessAndDebuggerNotification (Task task)
Microsoft.AppCenter.Storage.Storage
<FlushQueueAsync>d__34.MoveNext ()

### Details
What is your app platform (Xamarin.Android or Xamarin.iOS or UWP)?
WPF
If using Xamarin.Forms or if using portable/shared code to call our SDK APIs, are you using shared project, PCL code or .NET standard code for the application? Which .NET standard version or which PCL profile?
.N/A
Which SDK version are you using?
3.2.1
Which OS version did you experience the issue on?
Windows 10.0.17763
What device version did you see this error on? Were you using an emulator or a physical device?
VMware Virtual Platform
What third party libraries are you using?
N/A
Please enable verbose logging for your app using AppCenter.LogLevel = LogLevel.Verbose before your call to AppCenter.Start(...) and include the logs here: Cannot get the logs as it is happening on a client.

**Comment by annakocheshkova** [comment]
Hi there! Could you include a full raw json report (You can download it at the report page - upper right corner)? Thanks.

**Comment by tomaustin700** [comment]
Hi @annakocheshkova, I have attached the json report.

Thanks.
 
[report-2518069554637449999-53f77150-5da6-4183-b4a9-135700d6ada0.zip](https://github.com/microsoft/appcenter-sdk-dotnet/files/4967774/report-2518069554637449999-53f77150-5da6-4183-b4a9-135700d6ada0.zip)

**Comment by annakocheshkova** [comment]
@tomaustin700  is it possible that one of your clients starts the app from 2 places?
If so, this could happen.

**Comment by tomaustin700** [comment]
@AnastasiaKubova They could be using remote desktop services but I'm not 100% to be honest. Is there anything that we/you can do to help in that scenario? Thanks

**Comment by annakocheshkova** [comment]
I suppose database access settings might be causing the issue here. Could you check, do you receive those error reports from some specific customer all the time? Or the logs coming from his device show that the issue does not persist?

**Comment by tomaustin700** [comment]
We've had this come from separate users but only once per user.

**Comment by AnastasiaKubova** [comment]
This error usually happens when one app tries to open the connection to the database twice or two apps open the same database. The connection to the database in the App Center SDK created once after `AppCenter.Start`. If you are sure that the application could not be launch twice then could you please share additional information about this issue (reproducing steps or verbose logs) because this issue is not easy to reproduce without additional information?
Thanks

**Comment by tomaustin700** [comment]
@AnastasiaKubova Some of our clients deploy using remote desktop services so there is a possibility that the application could be launched twice. Is there anything we can do to mitigate this? We are not able to reproduce it locally and due to it happening with our clients we can't get verbose logs.

**Comment by AnastasiaKubova** [comment]
When I talk about launch the application twice, I mean, for example, launching the application via VS and `.exe` at the same time (in this case these apps have the same install id and the same database location).
In this case the only way to avoid this is to make sure that the same application hasn't been launched twice at the same time. 
I'm not sure that it might be happening due to the remote desktop connection.
I will label this issue as a bug.

**Comment by AnastasiaKubova** [comment]
The fix has been [released](https://github.com/microsoft/appcenter-sdk-dotnet/releases/tag/3.4.1).
Please update App Center SDK to the latest version and let me know if this error will appear again.

**Comment by Vitch612** [comment]
This issue is still happening in a WPF app running on .Net framework 4.8 with the following packages:

Microsoft.AppCenter 5.0.2
Microsoft.AppCenter.Analytics 5.0.2
Microsoft.AppCenter.Crashes 5.0.2
SQLitePCLRaw.bundle_green 2.1.5
SQLitePCLRaw.core 2.1.5
SQLitePCLRaw.lib.e_sqlite3 2.1.5
SQLitePCLRaw.provider.dynamic_cdecl 2.1.5

The solution is to manually add the dlls from SQLitePCLRaw.lib.e_sqlite3 and SQLitePCLRaw.bundle_green to the solution

**Comment by MikhailSuendukov** [comment]
@Vitch612 Hello and thank you for contacting us, please add information on how often this error occurs and what are the steps to reproduce. Also, could you please confirm if your bug is a duplicate of https://github.com/microsoft/appcenter-sdk-dotnet/issues/1740.

**Comment by Vitch612** [comment]
Hello @MikhailSuendukov, it is not a duplicate of #1740, my context is windows only. For additional context I am using Visual Studio 2019 community edition on a windows 7SP1.

I came accross this post by googling this specific line from console:
`Microsoft.AppCenter.Storage.StorageException in <>c__DisplayClass31_0`1.b__0 ()`

The steps to reproduce are to install the nuget packages mentionned in my earlier post. Both nuget configurations PackageReference and packages.config produce errors. [#1716](https://github.com/microsoft/appcenter-sdk-dotnet/issues/1716) is similar and recommends migrating to PackageReference, I have tried it but it did not solve the issue (though there was some improvement as the SQLitePCLRaw.lib.e_sqlite3 dlls were copied to the executable folder which was not happening in packages.config mode).

The main issue lies in the SQLitePCL packages and it seems it has been a known issue for a long time. Many threads similar to this one reach that same conclusion. 

Another thread with the same core issue: [#1452](https://github.com/microsoft/appcenter-sdk-dotnet/issues/1452)
And some duplicates of 1716 [#1722](https://github.com/microsoft/appcenter-sdk-dotnet/issues/1722)  [1583](https://github.com/microsoft/appcenter-sdk-dotnet/issues/1583)

Interesting because someone mentions having similar issues with .Net core 6 (I was under the impression this issue was exclusive to .Net framework): [#1583](https://github.com/microsoft/appcenter-sdk-dotnet/issues/1583)

Also of interest it seems like using this sqlite set of packages has been problematic all along to have generated this discussion: [#1354](https://github.com/microsoft/appcenter-sdk-dotnet/issues/1354)

Goinger further back in time: [#1313](https://github.com/microsoft/appcenter-sdk-dotnet/issues/1313)

I order to get it to work I have edited packages.config to modify as follows:
```
<package id="SQLitePCLRaw.bundle_green" version="2.1.5" targetFramework="net48" ExcludeAssets="All" />
<package id="SQLitePCLRaw.lib.e_sqlite3" version="2.1.5" targetFramework="net48" ExcludeAssets="All"/>
```
Then i manually added the runtimes folder from package SQLitePCLRaw.lib.e_sqlite3 to the project and specified the dlls should be copied. (having the folder structure reproduced seems to be important as simply copying the dll for my architecture to the project folder did not suffice)

I have also added the SQLitePCLRaw.batteries_v2.dll from SQLitePCLRaw.bundle_green.2.1.5 to the project had had it copied to the executable folder.

**Comment by MikhailSuendukov** [comment]
Are you still encountering this issue even after starting a new project from scratch, migrating to PackageReference, and then adding NuGet packages? It's quite an unusual error. I suspect there might be an issue with the cache being saved and reused.

**Comment by Vitch612** [comment]
I could not afford to start a new project on the app i wanted to add appcenter to,(old project with a lot of customizations) so i did not try it. There seems to be some confusion here, the issue i was having is solved, my initial post is not meant as a request for assistance but only to share how it was solved as i have spent a lot of time reading similar posts to manage to fix it. 

Following your last post I have tried a to add the appcenter packages and dependencies to a newly created project (same environment, same type of project) and it worked fine without needing any fix. So indeed the issue can't be reproduced at will.
